### PR TITLE
Add support EL for keys in formParamMap

### PR DESCRIPTION
**Pull request by Amerousful**
https://github.com/gatling/gatling/issues/3992

### CODE REVIEW

**Review by slandelle — CHANGES_REQUESTED** [review]
Hi,

Thanks for contributing. Before we can review, could you please run `sbt "compile;test:compile"` so that your code gets properly formatted and push again so that CI can verify your changes?

**Comment by Amerousful** [comment]
Hi. Okay, I will do this.

File: gatling-http/src/main/scala/io/gatling/http/request/builder/HttpRequestBuilder.scala [hunk 1]
```diff
@@ -128,9 +128,9 @@ final case class HttpRequestBuilder(commonAttributes: CommonAttributes, httpAttr
   def multivaluedFormParam(key: Expression[String], values: Expression[Seq[Any]]): HttpRequestBuilder = formParam(MultivaluedParam(key, values))
 
   def formParamSeq(seq: Seq[(String, Any)]): HttpRequestBuilder = formParamSeq(seq2SeqExpression(seq))
-  def formParamSeq(seq: Expression[Seq[(String, Any)]]): HttpRequestBuilder = formParam(ParamSeq(seq))
+  def formParamSeq(seq: Expression[Seq[(Any, Any)]]): HttpRequestBuilder = formParam(ParamSeq(seq))
 
-  def formParamMap(map: Map[String, Any]): HttpRequestBuilder = formParamSeq(map2SeqExpression(map))
+  def formParamMap(map: Map[String, String]): HttpRequestBuilder = formParamSeq(map2SeqExpression(map))
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/request/builder/HttpRequestBuilder.scala:133]: This is wrong: you're losing the ability to pass `formParamMap(Map("foo" -> 1))`.

File: gatling-http/src/main/scala/io/gatling/http/request/builder/HttpRequestBuilder.scala [hunk 2]
```diff
@@ -128,9 +128,9 @@ final case class HttpRequestBuilder(commonAttributes: CommonAttributes, httpAttr
   def multivaluedFormParam(key: Expression[String], values: Expression[Seq[Any]]): HttpRequestBuilder = formParam(MultivaluedParam(key, values))
 
   def formParamSeq(seq: Seq[(String, Any)]): HttpRequestBuilder = formParamSeq(seq2SeqExpression(seq))
-  def formParamSeq(seq: Expression[Seq[(String, Any)]]): HttpRequestBuilder = formParam(ParamSeq(seq))
+  def formParamSeq(seq: Expression[Seq[(Any, Any)]]): HttpRequestBuilder = formParam(ParamSeq(seq))
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/request/builder/HttpRequestBuilder.scala:131]: What's the use case for allowing to pass something else that a String as the key?
We mostly want to be able to pass `formParamSeq(Seq("${foo}" -> 1))` (an EL String key) while still being able to pass `formParamSeq(Seq("bar" -> 2))` (a regular String key).

File: gatling-http/src/main/scala/io/gatling/http/request/builder/HttpParam.scala [hunk 3]
```diff
@@ -21,5 +21,5 @@ import io.gatling.core.session.Expression
 sealed trait HttpParam
 final case class SimpleParam(key: Expression[String], value: Expression[Any]) extends HttpParam
 final case class MultivaluedParam(key: Expression[String], values: Expression[Seq[Any]]) extends HttpParam
-final case class ParamSeq(seq: Expression[Seq[(String, Any)]]) extends HttpParam
+final case class ParamSeq(seq: Expression[Seq[(Any, Any)]]) extends HttpParam
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/request/builder/HttpParam.scala:24]: Why? While values can be considered as not Strings (eg String representations of numbers), keys are always Strings.

File: gatling-core/src/main/scala/io/gatling/core/session/package.scala [hunk 4]
```diff
@@ -72,26 +78,22 @@ package object session {
...
-  def map2SeqExpression(map: Map[String, Any]): Expression[Seq[(String, Any)]] = {
-    val elValues: Map[String, Expression[Any]] = map.view
-      .mapValues {
-        case s: String => s.el[Any]
-        case v         => v.expressionSuccess
-      }
-      .to(Map)
-
-    resolveIterable(elValues)
+  def map2SeqExpression(map: Map[String, String]): Expression[Seq[(Any, Any)]] = {
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/session/package.scala:96]: See below, I think return type should be Expression[Seq[(String, Any)]]

File: gatling-core/src/main/scala/io/gatling/core/session/package.scala [hunk 5]
```diff
@@ -72,26 +78,22 @@ package object session {
     resolveRec(_, iterable.iterator, Nil)
   }
 
-  def seq2SeqExpression(seq: Seq[(String, Any)]): Expression[Seq[(String, Any)]] = {
-    val elValues: Seq[(String, Expression[Any])] = seq.map { case (key, value) =>
-      val elValue = value match {
-        case s: String => s.el[Any]
-        case v         => v.expressionSuccess
+  def seq2SeqExpression(seq: Seq[(Any, Any)]): Expression[Seq[(Any, Any)]] = {
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/session/package.scala:81]: See below, I think it should be seq2SeqExpression(seq: Seq[(String, Any)]): Expression[Seq[(String, Any)]]

File: gatling-http/src/test/scala/io/gatling/http/request/builder/HttpRequestBuilderSpec.scala [hunk 6]
```diff
@@ -92,6 +92,28 @@ class HttpRequestBuilderSpec extends BaseSpec with ValidationValues with EmptySe
...
+      "foo" -> "FOO",
+      "baz" -> "BAZ"
+    )
+
+    val session = sessionBase.setAll(form)
+
+    val params = httpRequestDef(
+      _.formParamMap(
+        Map(
+          "${foo}" -> "${baz}"
```
Comment (slandelle) [inline gatling-http/src/test/scala/io/gatling/http/request/builder/HttpRequestBuilderSpec.scala:106]: Please also test the other cases:
* when the key is a simple non- ELString
* when the value a simple non- ELString
* when the value is not a String

**Review by slandelle — CHANGES_REQUESTED** [review]

File: gatling-http/src/main/scala/io/gatling/http/request/builder/HttpRequestBuilder.scala (hunk 2 above)
Comment (Amerousful) [reply gatling-http/src/main/scala/io/gatling/http/request/builder/HttpRequestBuilder.scala]: Thanks) I understood my mistakes!

File: gatling-http/src/main/scala/io/gatling/http/request/builder/RequestBuilder.scala [hunk 7]
```diff
@@ -121,7 +121,7 @@ abstract class RequestBuilder[B <: RequestBuilder[B]] {
   def queryParamSeq(seq: Seq[(String, Any)]): B = queryParamSeq(seq2SeqExpression(seq))
   def queryParamSeq(seq: Expression[Seq[(String, Any)]]): B = queryParam(ParamSeq(seq))
 
-  def queryParamMap(map: Map[String, Any]): B = queryParamSeq(map2SeqExpression(map))
+  def queryParamMap(map: Map[String, String]): B = queryParamSeq(map2SeqExpression(map))
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/request/builder/RequestBuilder.scala:124]: Should be queryParamMap(map: Map[String, Any])

File: gatling-http/src/main/scala/io/gatling/http/request/builder/package.scala [hunk 8]
```diff
@@ -84,7 +84,7 @@ package object builder {
           for {
             seq <- seq(session)
           } yield {
-            seq.foreach { case (key, value) => clientParams.add(new Param(key, value.toString)) }
+            seq.foreach { case (key, value) => clientParams.add(new Param(key.toString, value.toString)) }
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/request/builder/package.scala:87]: key has to be a String so this can be reverted

**Review by slandelle — CHANGES_REQUESTED** [review]

**Comment by slandelle** [comment]
Looking great, thanks!
Could you please sign our [CLA](https://docs.google.com/forms/d/1hbpxVqJ5hIYYJuOjOyOzub9fJFGFGvvfnB5w5LCWdUw)?

**Comment by Amerousful** [comment]
Thank you) I'm glad to participate in the contribution!
Yes, I have just completed the document.

**Comment by slandelle** [comment]
Squashed and picked. Thanks a bunch!
As this involves a breaking change, that's something we'll only be able to ship in 3.6.0, not 3.5.2.

**Pull request closed without merging**